subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java

```diff
@@ -17,27 +17,25 @@
 
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Optional;
 
 /**
  * A converter which infers the class names from the file name.
  */
 public class FileNameDerivingClassNameConverter implements SourceFileClassNameConverter {
     @Override
     public Collection<String> getClassNames(String sourceFileRelativePath) {
-        return Collections.singleton(findClassNameForRelativePath(sourceFileRelativePath));
+        if (sourceFileRelativePath.endsWith(".java")) {
```

Comment:
Just making it clear here that this class should never be used by the Groovy compiler - If it is accidentally used, the Groovy compiler will silently ignore all changes, because it will return the file name as the class name. This cost me quite a bit of time to track down, so I decided to leave this in even though it is now (correctly) never used by the Groovy compiler.